### ISSUE TITLE
C API uses errno 

### ISSUE THREAD

**Issue by jtsylve** [Velocidex/c-aff4#114]
The current C API returns -1 on failure and sets `errno` with a specific error value.  This behavior should be avoided outside of system calls [1].

I suggest that the C API instead return an AFF4Status value, but am interested in your thoughts before I submit a PR.

[1] https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Re-no-throw

**Comment by scudette** [comment]
Yeah I think you are right - there wasn't much thought given to the C api - most of the work was done on C++ and there just was no user of the c api at the time. It makes sense that as people use it the pain points will come out and we would need to evolve that. One of the pain points is multi-threading support in general i expect and a global errno is bad for that.

An AFF4Status may not be descriptive enough for many cases and maybe a full error object would make sense but the C++ API doesn't do that either so it may be another discussion.
